Improve log display

AlanGreene opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

<!-- Describe your changes here- ideally you can get that description straight from your descriptive commit message(s)! -->
Related to https://github.com/tektoncd/dashboard/issues/2306
Resolves https://github.com/tektoncd/dashboard/issues/3737
  
Add support for toggling display of timestamps in the logs, so the user no longer has to navigate away to the settings page to do this. This is available on both PipelineRun and TaskRun pages. We always request timestamps and just show / hide them depending on the user's preference. This is persisted to browser localStorage as with the toggle on the settings page.

Add support for detecting GitHub Actions workflow command-style log levels in log output. This provides an improved user experience as it allows for filtering logs to hide unwanted noise, e.g. debug logs, by default. We may consider allowing the format to be customised in a future release depending on user feedback.

Refactor the styles so `LogFormat` now correctly owns most of the styling of the log content, with `Log` only responsible for additional styling of the container.

Refactor use of the `LogsToolbar` component to allow for customisable use by third-party consumers of the Dashboard components. This means they can much more easily take advantage of the new features, such as toggling timestamps and log levels, without having to reimplement the menu and related code themselves.

Eliminate redundant use of `split` and `join` calls when processing the logs, improving performance. `LogFormat` now receives an array of log line objects, pre-parsed into the new structure with the `timestamp`, `level` (optional), and `message` fields. Where a multiline log is encountered, the timestamp of the first line is reused for subsequent lines in that log.

Fix issue where in some cases a blank line did not reserve vertical space, leading to cramped display of logs. Now each line is guaranteed to occupy a minimum height, ensuring blank lines output in the logs to aid in readability are preserved in the UI.

Update `FormattedDate` to add support for displaying seconds, as this is quite important in the log context. Default to `false` for this setting so existing date / timestamps in other parts of the UI are unaffected. The full raw timestamp as received in the logs in displayed in a tooltip on hover.

Update unit tests to reflect the new and changed components and behaviours.

Update common PipelineRun E2E to exercise the new log toolbar and validate the log content is rendered as expected.

Add new stories to cover the new functionality. Update existing stories to demonstrate use of the new functionality in context.

Update Carbon:
- resolve issue with Plex Mono font

  Some glyphs weren't included in the Plex version packaged with previous Carbon releases, resulting in broken formatting for some log content, e.g. using box characters to print tables.

  In `@carbon/react` 1.71.0 the Plex version has been updated, as well as changing how it's consumed. Instead of a single package with all of the font variants, they're now published as separate packages per font family. Add the `$use-per-family-plex` flag to our config to use these new packages. The custom `$font-path` is still required for compatibility with Vite.

- resolve issue with duplicate onChange events from MenuItemSelectable

- resolve issue with duplicate onChange events when clearing a ComboBox

Notes:
- colours of the log level badges are based on the colours of the Carbon `Tag` component, with their opacity reduced so they're not as intense due to the potentially large number of them that could be displayed in the logs. These all meet minimum colour contrast ratio required for WCAG 2.0 level AA (i.e. > 4.5:1).
- the default log level is 'info' if no log level is explicitly provided in the logs, however we only display the badge when the log level is explicitly set. This avoids unnecessary and unwanted noise / clutter in the logs when not using the new log format.
- highlight and hover state included to highlight log lines, aiding in consuming the content, especially with longer log lines where the log level badge may not be adjacent to the content being read. A future update to the log viewer will add support for line wrapping but this is out of scope for this particular change.

/kind feature
/kind design
/kind bug
/kind misc
/kind documentation

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (new features, significant UI changes, API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Add timestamp toggle and log level filtering to the log viewer.
The log viewer now supports GitHub Actions workflow command-style log levels in log output (e.g. `::warning::This is an important message…`), allowing for filtering logs to hide unwanted noise, e.g. debug logs, by default. If no log level is detected, the line is treated as `info` by default, but only lines with an explicitly set log level display the badge.

Supported commands are: `::error::`, `::warning::`, `::notice::`, `::info::`, and `::debug::`
```

AlanGreene commented:
/test all

AlanGreene commented:
E2E failed on import resources test when selecting the import namespace. Not seeing the same issue locally, rerunning to see if it persists on prow

/test pull-tekton-dashboard-integration-tests

AlanGreene commented:
Failed in the same place, will need to debug next week…

briangleeson commented on src/containers/LogsToolbar/LogsToolbar.jsx:
```diff
@@ -0,0 +1,54 @@
+  const { namespace } = taskRun.metadata;
+  const { podName } = taskRun.status;
+
+  const logURL = isUsingExternalLogs
```
What's the external logs URL use case?

AlanGreene replied:
It's for the functionality described by https://github.com/tektoncd/dashboard/blob/main/docs/walkthrough/walkthrough-logs.md

Where logs are archived to external storage on run completion, so they can still be retrieved after the pods have been removed from the cluster.

briangleeson reviewed (commented):
A few queries but overall looking good

briangleeson commented on src/containers/PipelineRun/PipelineRun.jsx:
```diff
@@ -49,21 +50,49 @@ import {
+    setLogLevelsState(levels => {
+      const newLevels = { ...levels, ...logLevel };
+      // if (!Object.values(newLevels).filter(Boolean).length) {
+      //   // TODO: logs - notification or allow?
```
Could we place a single log line stating that all log levels have been disabled?

briangleeson commented:
One thing I noticed at https://alangreene.github.io/dashboard/?path=/story/components-log--toolbar, the dropdown menu disappears every time you change a logLevel option. Would be handy to be keep it up if user wants to add/remove multiple levels

AlanGreene commented on src/containers/TaskRun/TaskRun.jsx:
```diff
@@ -56,15 +53,47 @@ import {
+    setLogLevelsState(levels => {
+      const newLevels = { ...levels, ...logLevel };
+      // if (!Object.values(newLevels).filter(Boolean).length) {
+      //   // TODO: logs - notification or allow?
```
Yes, pending discussion with design this week 🤞

AlanGreene commented on src/containers/PipelineRun/PipelineRun.jsx:
```diff
@@ -49,21 +50,49 @@ import {
 const { PIPELINE_TASK, RETRY, STEP, TASK_RUN_NAME, VIEW } = queryParamConstants;
 
-export /* istanbul ignore next */ function PipelineRunContainer() {
+export /* istanbul ignore next */ function PipelineRunContainer({
```
TaskRun has an `ignore file` at the top

AlanGreene commented:
> dropdown menu disappears every time you change a logLevel option. Would be handy to be keep it up if user wants to add/remove multiple levels

Yes, that's something we've already discussed with design. Other similar tools behave in this way, including the Google Chrome dev tools. We may revisit based on user feedback and provide a richer popover similar to the Mezmo log level selection or something else entirely.

AlanGreene commented:
Forgot to add an update this morning. The Carbon issue causing the E2E failure is https://github.com/carbon-design-system/carbon/issues/18145, which explains why it wasn't reproduced locally in dev mode.

AlanGreene commented:
Fix for https://github.com/carbon-design-system/carbon/issues/18145 was released and we've updated to the relevant version in https://github.com/tektoncd/dashboard/pull/3825

Just the last outstanding design question about discoverability / displaying count of hidden lines to be resolved before merging.

AlanGreene commented:
/test all

AlanGreene commented:
Switched to new approach for the settings using a custom `Popover` component so the user can more easily make multiple changes. Also added a notice at the top of the logs when lines are hidden due to the currently selected log levels.

briangleeson commented on docs/logs.md:
```diff
@@ -0,0 +1,79 @@
+
+## Toolbar
+
+The toolbar diplayed in the log viewer includes a number of additional features, including:
```
Typo
```suggestion
The toolbar displayed in the log viewer includes a number of additional features, including:
```

AlanGreene replied:
🙈

briangleeson commented on docs/logs.md:
```diff
@@ -0,0 +1,79 @@
+
+The Dashboard will always request logs with timestamps from the Kubernetes pod logs API, and show / hide the timestamps in the log viewer based on the user's preference. This can be toggled from the settings menu in the toolbar at the top of the log viewer.
+
+The timestamps are localised based on users' browser settings, with the raw timestamp value received from the API provided as a tooltip on hover.
```
Was this intentional usage of plural `users'` ownership apostrophe? I think singular reads a bit better?
```suggestion
The timestamps are localised based on the user's browser settings, with the raw timestamp value received from the API provided as a tooltip on hover.
```

AlanGreene replied:
Yeah either works, and I notice I used the singular in the previous paragraph so will update here too. Thanks 👍

briangleeson reviewed (changes requested):
Looks good overall, thanks for the walkthrough. Just spotted some typos in the doc

briangleeson reviewed (approved):
/lgtm

👍

AlanGreene commented:
/retest
